src/hotspot/cpu/x86/gc/g1/g1BarrierSetAssembler_x86.cpp

```diff
@@ -160,6 +163,56 @@ void G1BarrierSetAssembler::load_at(MacroAssembler* masm, DecoratorSet decorator
   }
 }
 
+static void generate_queue_insertion(MacroAssembler* masm, ByteSize index_offset, ByteSize buffer_offset, Label& runtime,
+                                     const Register thread, const Register value, const Register temp) {
+  // This code assumes that buffer index is pointer sized.
+  STATIC_ASSERT(in_bytes(SATBMarkQueue::byte_width_of_index()) == sizeof(intptr_t));
+  // Can we store a value in the given thread's buffer?
+  // (The index field is typed as size_t.)
+  __ movptr(temp, Address(thread, in_bytes(index_offset)));   // temp := *(index address)
+  __ testptr(temp, temp);                                     // index == 0?
+  __ jcc(Assembler::zero, runtime);                           // jump to runtime if index == 0 (full buffer)
+  // The buffer is not full, store value into it.
+  __ subptr(temp, wordSize);                                  // temp := next index
+  __ movptr(Address(thread, in_bytes(index_offset)), temp);   // *(index address) := next index
+  __ addptr(temp, Address(thread, in_bytes(buffer_offset)));  // temp := buffer address + next index
+  __ movptr(Address(temp, 0), value);                         // *(buffer address + next index) := value
+}
+
+static void generate_pre_barrier_fast_path(MacroAssembler* masm,
+                                           const Register thread) {
+  Address in_progress(thread, in_bytes(G1ThreadLocalData::satb_mark_queue_active_offset()));
+  // Is marking active?
+  if (in_bytes(SATBMarkQueue::byte_width_of_active()) == 4) {
+    __ cmpl(in_progress, 0);
+  } else {
+    assert(in_bytes(SATBMarkQueue::byte_width_of_active()) == 1, "Assumption");
+    __ cmpb(in_progress, 0);
+  }
+}
+
+static void generate_pre_barrier_slow_path(MacroAssembler* masm,
+                                           const Register obj,
+                                           const Register pre_val,
+                                           const Register thread,
+                                           const Register tmp,
+                                           Label& done,
+                                           Label& runtime) {
+  // Do we need to load the previous value?
+  if (obj != noreg) {
+    __ load_heap_oop(pre_val, Address(obj, 0), noreg, noreg, AS_RAW);
```

Comment:
How do we handle implicit null checks for which `obj` is null? Note that we may expect the store instruction to trigger SIGSEGV. Does it work correctly if we trigger the SIGSEGV, here in the pre barrier?

Reply:
Thanks for figuring it out! Makes sense.

Reply:
Added the assertion in commit 554de779.